# Recommendations with IBM

In this notebook, you will be putting your recommendation skills to use on real data from the IBM Watson Studio platform. 


You may either submit your notebook through the workspace here, or you may work from your local machine and submit through the next page.  Either way assure that your code passes the project [RUBRIC](https://review.udacity.com/#!/rubrics/2322/view).  **Please save regularly.**

By following the table of contents, you will build out a number of different methods for making recommendations that can be used for different situations. 


## Table of Contents

I. [Exploratory Data Analysis](#Exploratory-Data-Analysis)<br>
II. [Rank Based Recommendations](#Rank)<br>
III. [User-User Based Collaborative Filtering](#User-User)<br>
IV. [Content Based Recommendations (EXTRA - NOT REQUIRED)](#Content-Recs)<br>
V. [Matrix Factorization](#Matrix-Fact)<br>
VI. [Extras & Concluding](#conclusions)

At the end of the notebook, you will find directions for how to submit your work.  Let's get started by importing the necessary libraries and reading in the data.

In [1174]:
import random

In [775]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import project_tests as t
import pickle

%matplotlib inline

df = pd.read_csv('data/user-item-interactions.csv')
df_content = pd.read_csv('data/articles_community.csv')
del df['Unnamed: 0']
del df_content['Unnamed: 0']

# Show df to get an idea of the data
df.head()

article_id                                              title  \
0      1430.0  using pixiedust for fast, flexible, and easier...   
1      1314.0       healthcare python streaming application demo   
2      1429.0         use deep learning for image classification   
3      1338.0          ml optimization using cognitive assistant   
4      1276.0          deploy your python model as a restful api   

                                      email  
0  ef5f11f77ba020cd36e1105a00ab868bbdbf7fe7  
1  083cbdfa93c8444beaa4c5f5e0f5f9198e4f9e0b  
2  b96a4f2e92d8572034b1e9b28f9ac673765cd074  
3  06485706b34a5c9bf2a0ecdac41daf7e7654ceb7  
4  f01220c46fc92c6e6b161b1849de11faacd7ccb2

### <a class="anchor" id="Exploratory-Data-Analysis">Part I : Exploratory Data Analysis</a>

Use the dictionary and cells below to provide some insight into the descriptive statistics of the data.

`1.` What is the distribution of how many articles a user interacts with in the dataset?  Provide a visual and descriptive statistics to assist with giving a look at the number of times each user interacts with an article.  

In [779]:
# 50% of individuals interact with ____ number of articles or fewer.

median_val = df.groupby('email').size().sort_values(ascending=True).median()

median_val

3.0

In [780]:
# The maximum number of user-article interactions by any 1 user is ______. ?

max_views_by_user = df.groupby('email').size().sort_values(ascending=False).head(1).values[0]

max_views_by_user

364

`2.` Explore and remove duplicate articles from the **df_content** dataframe.  

In [781]:
df_content['article_id'].nunique()

1051

In [782]:
# Find and explore duplicate articles

dups = df_content.groupby('article_id').size().sort_values(ascending=False) # duplicates of article id
dups_idx = dups[dups > 1].index # duplicates index

df_content[df_content['article_id'].isin(dups_idx)].sort_values(by='article_id') # retrieve duplicated content

doc_body  \
50   Follow Sign in / Sign up Home About Insight Da...   
365  Follow Sign in / Sign up Home About Insight Da...   
221  * United States\r\n\r\nIBM® * Site map\r\n\r\n...   
692  Homepage Follow Sign in / Sign up Homepage * H...   
232  Homepage Follow Sign in Get started Homepage *...   
971  Homepage Follow Sign in Get started * Home\r\n...   
399  Homepage Follow Sign in Get started * Home\r\n...   
761  Homepage Follow Sign in Get started Homepage *...   
578  This video shows you how to construct queries ...   
970  This video shows you how to construct queries ...   

                                       doc_description  \
50                        Community Detection at Scale   
365  During the seven-week Insight Data Engineering...   
221  When used to make sense of huge amounts of con...   
692  One of the earliest documented catalogs was co...   
232  If you are like most data scientists, you are ...   
971  If you are like most data scientists, you are ...   
399  Today’s world of data science leverages data f...   
761  Today’s world of data science leverages data f...   
578  This video shows you how to construct queries ...   
970  This video shows you how to construct queries ...   

                                         doc_full_name doc_status  article_id  
50                        Graph-based machine learning       Live          50  
365                       Graph-based machine learning       Live          50  
221  How smart catalogs can turn the big data flood...       Live         221  
692  How smart catalogs can turn the big data flood...       Live         221  
232  Self-service data preparation with IBM Data Re...       Live         232  
971  Self-service data preparation with IBM Data Re...       Live         232  
399  Using Apache Spark as a parallel processing fr...       Live         398  
761  Using Apache Spark as a parallel processing fr...       Live         398  
578                              Use the Primary Index       Live         577  
970                              Use the Primary Index       Live         577

In [783]:
# Remove any rows that have the same article_id - only keep the first

df_content = df_content.drop_duplicates('article_id', keep='first')

`3.` Use the cells below to find:

**a.** The number of unique articles that have an interaction with a user.  
**b.** The number of unique articles in the dataset (whether they have any interactions or not).<br>
**c.** The number of unique users in the dataset. (excluding null values) <br>
**d.** The number of user-article interactions in the dataset.

In [785]:
unique_articles = df.article_id.nunique() # The number of unique articles that have at least one interaction
unique_articles

714

In [786]:
total_articles = df_content.article_id.nunique() # The number of unique articles on the IBM platform
total_articles

1051

In [787]:
unique_users = df.email.nunique() # The number of unique users
unique_users


5148

In [788]:
user_article_interactions = df.shape[0] # The number of user-article interactions

user_article_interactions

45993

`4.` Use the cells below to find the most viewed **article_id**, as well as how often it was viewed.  After talking to the company leaders, the `email_mapper` function was deemed a reasonable way to map users to ids.  There were a small number of null values, and it was found that all of these null values likely belonged to a single user (which is how they are stored using the function below).

In [789]:
most_viewed_ds = df.groupby('article_id').size().sort_values(ascending=False).head(1)

In [790]:
# The most viewed article in the dataset as a string with one value following the decimal 
most_viewed_article_id = str(most_viewed_ds.index[0])
most_viewed_article_id

'1429.0'

In [791]:
# The most viewed article in the dataset was viewed how many times?

max_views = most_viewed_ds.values[0]
max_views

937

In [792]:
## No need to change the code here - this will be helpful for later parts of the notebook
# Run this cell to map the user email to a user_id column and remove the email column

def email_mapper():
    coded_dict = dict()
    cter = 1
    email_encoded = []
    
    for val in df['email']:
        if val not in coded_dict:
            coded_dict[val] = cter
            cter+=1
        
        email_encoded.append(coded_dict[val])
    return email_encoded

email_encoded = email_mapper()
del df['email']
df['user_id'] = email_encoded

# show header
df.head()

article_id                                              title  user_id
0      1430.0  using pixiedust for fast, flexible, and easier...        1
1      1314.0       healthcare python streaming application demo        2
2      1429.0         use deep learning for image classification        3
3      1338.0          ml optimization using cognitive assistant        4
4      1276.0          deploy your python model as a restful api        5

In [793]:
## If you stored all your results in the variable names above, 
## you shouldn't need to change anything in this cell

sol_1_dict = {
    '`50% of individuals have _____ or fewer interactions.`': median_val, 
    '`The total number of user-article interactions in the dataset is ______.`': user_article_interactions, 
    '`The maximum number of user-article interactions by any 1 user is ______.`': max_views_by_user,
    '`The most viewed article in the dataset was viewed _____ times.`': max_views, 
    '`The article_id of the most viewed article is ______.`': most_viewed_article_id, 
    '`The number of unique articles that have at least 1 rating ______.`': unique_articles, 
    '`The number of unique users in the dataset is ______`': unique_users, 
    '`The number of unique articles on the IBM platform`': total_articles 
}

# Test your dictionary against the solution
t.sol_1_test(sol_1_dict)

It looks like you have everything right here! Nice job!


### <a class="anchor" id="Rank">Part II: Rank-Based Recommendations</a>

Unlike in the earlier lessons, we don't actually have ratings for whether a user liked an article or not.  We only know that a user has interacted with an article.  In these cases, the popularity of an article can really only be based on how often an article was interacted with.

`1.` Fill in the function below to return the **n** top articles ordered with most interactions as the top. Test your function using the tests below.

In [794]:
def get_top_articles(n, df=df):
    '''
    INPUT:
    n - (int) the number of top articles to return
    df - (pandas dataframe) df as defined at the top of the notebook 
    
    OUTPUT:
    top_articles - (list) A list of the top 'n' article titles 
    
    '''
    # Your code here
    article_ids_ds = df.groupby('article_id').size().sort_values(ascending=False).head(n)
    ids = article_ids_ds.index
    
    top_articles = []
    for i in ids:
        top_articles.append(df[df['article_id'] == i].head(1).title.values[0])
    
    
    
    return top_articles # Return the top article titles from df (not df_content)

In [795]:
def get_top_article_ids(n, df=df):
    '''
    INPUT:
    n - (int) the number of top articles to return
    df - (pandas dataframe) df as defined at the top of the notebook 
    
    OUTPUT:
    top_articles_ids - (list) A list of the top 'n' article titles 
    
    '''
    # Your code here
    article_ids_ds = df.groupby('article_id').size().sort_values(ascending=False).head(n)
    ids = article_ids_ds.index
    top_articles_ids = list(ids)
 
    return top_articles_ids # Return the top article ids

In [799]:
print(get_top_articles(10))
print(get_top_article_ids(10))

['use deep learning for image classification', 'insights from new york car accident reports', 'visualize car data with brunel', 'use xgboost, scikit-learn & ibm watson machine learning apis', 'predicting churn with the spss random tree algorithm', 'healthcare python streaming application demo', 'finding optimal locations of new store using decision optimization', 'apache spark lab, part 1: basic concepts', 'analyze energy consumption in buildings', 'gosales transactions for logistic regression model']
[1429.0, 1330.0, 1431.0, 1427.0, 1364.0, 1314.0, 1293.0, 1170.0, 1162.0, 1304.0]


In [800]:
# Test your function by returning the top 5, 10, and 20 articles
top_5 = get_top_articles(5)
top_10 = get_top_articles(10)
top_20 = get_top_articles(20)

# Test each of your three lists from above
t.sol_2_test(get_top_articles)

Your top_5 looks like the solution list! Nice job.
Your top_10 looks like the solution list! Nice job.
Your top_20 looks like the solution list! Nice job.


### <a class="anchor" id="User-User">Part III: User-User Based Collaborative Filtering</a>


`1.` Use the function below to reformat the **df** dataframe to be shaped with users as the rows and articles as the columns.  

* Each **user** should only appear in each **row** once.


* Each **article** should only show up in one **column**.  


* **If a user has interacted with an article, then place a 1 where the user-row meets for that article-column**.  It does not matter how many times a user has interacted with the article, all entries where a user has interacted with an article should be a 1.  


* **If a user has not interacted with an item, then place a zero where the user-row meets for that article-column**. 

Use the tests to make sure the basic structure of your matrix matches what is expected by the solution.

In [532]:
# create the user-article matrix with 1's and 0's

def create_user_item_matrix(df):
    '''
    INPUT:
    df - pandas dataframe with article_id, title, user_id columns
    
    OUTPUT:
    user_item - user item matrix 
    
    Description:
    Return a matrix with user ids as rows and article ids on the columns with 1 values where a user interacted with 
    an article and a 0 otherwise
    '''
    # Fill in the function here
    df['title'] = 1
    df = df[['user_id', 'article_id', 'title']]
    
    user_item = pd.pivot_table(df, index='user_id', columns='article_id', values='title', fill_value=0)
    
    return user_item # return the user_item matrix 

user_item = create_user_item_matrix(df)

In [533]:
user_item

article_id  0.0     2.0     4.0     8.0     9.0     12.0    14.0    15.0    \
user_id                                                                      
1                0       0       0       0       0       0       0       0   
2                0       0       0       0       0       0       0       0   
3                0       0       0       0       0       1       0       0   
4                0       0       0       0       0       0       0       0   
5                0       0       0       0       0       0       0       0   
...            ...     ...     ...     ...     ...     ...     ...     ...   
5145             0       0       0       0       0       0       0       0   
5146             0       0       0       0       0       0       0       0   
5147             0       0       0       0       0       0       0       0   
5148             0       0       0       0       0       0       0       0   
5149             0       0       0       0       0       0       0       0   

article_id  16.0    18.0    ...  1434.0  1435.0  1436.0  1437.0  1439.0  \
user_id                     ...                                           
1                0       0  ...       0       0       1       0       1   
2                0       0  ...       0       0       0       0       0   
3                0       0  ...       0       0       1       0       0   
4                0       0  ...       0       0       0       0       0   
5                0       0  ...       0       0       0       0       0   
...            ...     ...  ...     ...     ...     ...     ...     ...   
5145             0       0  ...       0       0       0       0       0   
5146             0       0  ...       0       0       0       0       0   
5147             0       0  ...       0       0       0       0       0   
5148             0       0  ...       0       0       0       0       0   
5149             1       0  ...       0       0       0       0       0   

article_id  1440.0  1441.0  1442.0  1443.0  1444.0  
user_id                                             
1                0       0       0       0       0  
2                0       0       0       0       0  
3                0       0       0       0       0  
4                0       0       0       0       0  
5                0       0       0       0       0  
...            ...     ...     ...     ...     ...  
5145             0       0       0       0       0  
5146             0       0       0       0       0  
5147             0       0       0       0       0  
5148             0       0       0       0       0  
5149             0       0       0       0       0  

[5149 rows x 714 columns]

In [534]:
## Tests: You should just need to run this cell.  Don't change the code.
assert user_item.shape[0] == 5149, "Oops!  The number of users in the user-article matrix doesn't look right."
assert user_item.shape[1] == 714, "Oops!  The number of articles in the user-article matrix doesn't look right."
assert user_item.sum(axis=1)[1] == 36, "Oops!  The number of articles seen by user 1 doesn't look right."
print("You have passed our quick tests!  Please proceed!")

You have passed our quick tests!  Please proceed!


2. Complete the function below which should take a user_id and provide an ordered list of the most similar users to that user (from most similar to least similar). The returned result should not contain the provided user_id, as we know that each user is similar to him/herself. Because the results for each user here are binary, it (perhaps) makes sense to compute similarity as the dot product of two users.

Use the tests to test your function.

In [873]:
import time

In [697]:
# experiment with pandas inbuilt corr

user_item.loc[(1,2), :].transpose().corr()

user_id         1         2
user_id                    
1        1.000000  0.119027
2        0.119027  1.000000

2. Complete the function below which should take a user_id and provide an ordered list of the most similar users to that user (from most similar to least similar). The returned result should not contain the provided user_id, as we know that each user is similar to him/herself. **Because the results for each user here are binary, it (perhaps) makes sense to compute similarity as the dot product of two users.**

Use the tests to test your function.

In [1001]:
def find_similar_users(user_id, user_item=user_item):
    '''
    INPUT:
    user_id - (int) a user_id
    user_item - (pandas dataframe) matrix of users by articles: 
                1's when a user has interacted with an article, 0 otherwise
    
    OUTPUT:
    similar_users - (list) an ordered list where the closest users (largest dot product users)
                    are listed first
    
    Description:
    Computes the similarity of every pair of users based on the dot product
    Returns an ordered
    
    '''
    
    user_ids_all = user_item.index
    
    # a dictionary holds [user id : similarity] value pairs
    dot_product_dict = {}

    # compute similarity of each user to the provided user
    for i in user_ids_all:
        dot_product = np.dot(user_item.loc[user_id:user_id,:], user_item.loc[i:i,:].transpose())
        dot_product_dict[i] = dot_product[0][0]
        
    # sort by highest (most similar), also drop against self    
    ds = pd.Series(dot_product_dict.values(), index=dot_product_dict.keys()).drop(index=user_id).sort_values(ascending=False)
    
    # extract only index as list of ids. 
    most_similar_users = list(ds.index)
       
    return most_similar_users # return a list of the users in order from most to least similar

In [1004]:
# Do a spot check of your function
print("The 10 most similar users to user 1 are: {}".format(find_similar_users(1)[:10]))
print("The 5 most similar users to user 3933 are: {}".format(find_similar_users(3933)[:5]))
print("The 3 most similar users to user 46 are: {}".format(find_similar_users(46)[:3]))

The 10 most similar users to user 1 are: [3933, 23, 3782, 203, 4459, 3870, 131, 46, 4201, 395]
The 5 most similar users to user 3933 are: [1, 23, 3782, 4459, 203]
The 3 most similar users to user 46 are: [4201, 23, 3782]


`3.` Now that you have a function that provides the most similar users to each user, you will want to use these users to find articles you can recommend.  Complete the functions below to return the articles you would recommend to each user. 

In [1053]:
def get_article_names(article_ids, df=df):
    '''
    INPUT:
    article_ids - (list) a list of article ids
    df - (pandas dataframe) df as defined at the top of the notebook
    
    OUTPUT:
    article_names - (list) a list of article names associated with the list of article ids 
                    (this is identified by the title column)
    '''
    # Your code here
    
    article_names = []
    
    for id in article_ids:

        name = df[df.article_id == id].title.head(1).values[0]

        article_names.append(name)
    
    return article_names # Return the article names associated with list of article ids


a_ids = [1430.0, 1314.0, 1276.0, 233.0]
get_article_names(a_ids)

['using pixiedust for fast, flexible, and easier data analysis and experimentation',
 'healthcare python streaming application demo',
 'deploy your python model as a restful api',
 'bayesian nonparametric models – stats and bots']

In [1056]:
def get_user_articles(user_id, user_item=user_item):
    '''
    INPUT:
    user_id - (int) a user id
    user_item - (pandas dataframe) matrix of users by articles: 
                1's when a user has interacted with an article, 0 otherwise
    
    OUTPUT:
    article_ids - (list) a list of the article ids seen by the user
    article_names - (list) a list of article names associated with the list of article ids 
                    (this is identified by the doc_full_name column in df_content)
    
    Description:
    Provides a list of the article_ids and article titles that have been seen by a user
    '''
    # Your code here
    
    
    article_ids = list(user_item.loc[user_id][user_item.loc[user_id] > 0].index)
    article_names = get_article_names(article_ids)
    
    return article_ids, article_names # return the ids and names


get_user_articles(2)

([1024.0, 1176.0, 1305.0, 1314.0, 1422.0, 1427.0],
 ['using deep learning to reconstruct high-resolution audio',
  'build a python app on the streaming analytics service',
  'gosales transactions for naive bayes model',
  'healthcare python streaming application demo',
  'use r dataframes & ibm watson natural language understanding',
  'use xgboost, scikit-learn & ibm watson machine learning apis'])

In [1185]:
def user_user_recs(user_id, m=10):
    '''
    INPUT:
    user_id - (int) a user id
    m - (int) the number of recommendations you want for the user
    
    OUTPUT:
    recs - (list) a list of recommendations for the user
    
    Description:
    Loops through the users based on closeness to the input user_id
    For each user - finds articles the user hasn't seen before and provides them as recs
    Does this until m recommendations are found
    
    Notes:
    Users who are the same closeness are chosen arbitrarily as the 'next' user
    
    For the user where the number of recommended articles starts below m 
    and ends exceeding m, the last items are chosen arbitrarily
    
    '''
    # Your code here
    
    
    # get all similar user ids for the targeted user
    similar_uids = find_similar_users(user_id)


    # get all article ids of the targeted user
    article_ids_target_user = get_user_articles(user_id)[0]
    # print(f"[article_ids_target_user]:\n {article_ids_target_user} \n")

    
    # a list contain unseen articles to recommend
    recs = []

    for uid in similar_uids:

        # print(f"\n\n\n[number of recs]: {len(recs)}\n")
        if len(recs) == m:
            # print(f"Number of recs reaches threadhold. Enough. Stop")
            break


        #print(f"[similar user id]: {uid}")

        # get this uid's article ids, and arbitrarily shuffle
        article_ids_similar_user = get_user_articles(uid)[0]

        # compute the differences of articles seen between the this user and targeted user
        # subtraction's order matters
        set_diff = list(set(article_ids_similar_user) - set(article_ids_target_user))
        # print(f"[set_diff]:\n {set_diff} \n")
        
        # make a shuffle for arbitraily chocies from the diff set
        random.shuffle(set_diff)

        # add the differences of article ids to recs [], append only unique (no duplicate)
        for i in set_diff:
            if i not in recs and len(recs) < m:
                recs.append(i)
                # print(f"[id] {i} appended")

    return recs # return your recommendations for this user_id 

In [1146]:
# Check Results
get_article_names(user_user_recs(1, 10)) # Return 10 recommendations for user 1

['data tidying in data science experience',
 'this week in data science (april 18, 2017)',
 'shaping data with ibm data refinery',
 'data science platforms are on the rise and ibm is leading the way',
 'timeseries data analysis of iot events by using jupyter notebook',
 'got zip code data? prep it for analytics. – ibm watson data lab – medium',
 'higher-order logistic regression for large datasets',
 'using machine learning to predict parking difficulty',
 'a tensorflow regression model to predict house values',
 'deep forest: towards an alternative to deep neural networks']

In [1168]:
# BUT THE ANSWER WANT ARTICLE IDS TO BE STRING RATHER THAN NUMBER. However according to dataset, 
# they are all float. In the case, I just use its default astype for argument. 

df.article_id.dtype

dtype('float64')

In [1163]:
# Test your functions here
assert set(get_article_names([1024.0, 1176.0, 1305.0, 1314.0, 1422.0, 1427.0])) == set(['using deep learning to reconstruct high-resolution audio', 'build a python app on the streaming analytics service', 'gosales transactions for naive bayes model', 'healthcare python streaming application demo', 'use r dataframes & ibm watson natural language understanding', 'use xgboost, scikit-learn & ibm watson machine learning apis']), "Oops! Your the get_article_names function doesn't work quite how we expect."
assert set(get_article_names([1320.0, 232.0, 844.0])) == set(['housing (2015): united states demographic measures','self-service data preparation with ibm data refinery','use the cloudant-spark connector in python notebook']), "Oops! Your the get_article_names function doesn't work quite how we expect."
assert set(get_user_articles(20)[0]) == set([1320.0, 232.0, 844.0])
assert set(get_user_articles(20)[1]) == set(['housing (2015): united states demographic measures', 'self-service data preparation with ibm data refinery','use the cloudant-spark connector in python notebook'])
assert set(get_user_articles(2)[0]) == set([1024.0, 1176.0, 1305.0, 1314.0, 1422.0, 1427.0])
assert set(get_user_articles(2)[1]) == set(['using deep learning to reconstruct high-resolution audio', 'build a python app on the streaming analytics service', 'gosales transactions for naive bayes model', 'healthcare python streaming application demo', 'use r dataframes & ibm watson natural language understanding', 'use xgboost, scikit-learn & ibm watson machine learning apis'])
print("If this is all you see, you passed all of our tests!  Nice job!")

If this is all you see, you passed all of our tests!  Nice job!


`4.` Now we are going to improve the consistency of the **user_user_recs** function from above.  

* Instead of arbitrarily choosing when we obtain users who are all the same closeness to a given user - choose the users that have the most total article interactions before choosing those with fewer article interactions.


* Instead of arbitrarily choosing articles from the user where the number of recommended articles starts below m and ends exceeding m, choose articles with the articles with the most total interactions before choosing those with fewer total interactions. This ranking should be  what would be obtained from the **top_articles** function you wrote earlier.

In [1256]:
def get_top_sorted_users(user_id, df=df, user_item=user_item):
    '''
    INPUT:
    user_id - (int)
    df - (pandas dataframe) df as defined at the top of the notebook 
    user_item - (pandas dataframe) matrix of users by articles: 
            1's when a user has interacted with an article, 0 otherwise
    
            
    OUTPUT:
    neighbors_df - (pandas dataframe) a dataframe with:
                    neighbor_id - is a neighbor user_id
                    similarity - measure of the similarity of each user to the provided user_id
                    num_interactions - the number of articles viewed by the user - if a u
                    
    Other Details - sort the neighbors_df by the similarity and then by number of interactions where 
                    highest of each is higher in the dataframe
     
    '''
    # Your code here
    
    # get all neighbors ids
    nbh_ids = find_similar_users(user_id)
    
    
    # assemble a data matrix
    data_matrix = np.array([
    [
        x, # neighbor id
        np.dot(user_item.loc[user_id:user_id,:], user_item.loc[x:x,:].transpose())[0][0], # similarity score
        df[df.user_id == x].shape[0] # number of content interaction
    ] for x in nbh_ids])
    
    # make a dataframe
    neighbors_df = pd.DataFrame(data=data_matrix, 
                                columns=['neighbor_id', 'similarity', 'num_interactions'], 
                                index=data_matrix[:,0]).sort_values(by=['similarity', 'num_interactions'],
                                                                    ascending=[False, False])
    
    
    
    return neighbors_df # Return the dataframe specified in the doc_string

In [1309]:
def user_user_recs_part2(user_id, m=10):
    '''
    INPUT:
    user_id - (int) a user id
    m - (int) the number of recommendations you want for the user
    
    OUTPUT:
    recs - (list) a list of recommendations for the user by article id
    rec_names - (list) a list of recommendations for the user by article title
    
    Description:
    Loops through the users based on closeness to the input user_id
    For each user - finds articles the user hasn't seen before and provides them as recs
    Does this until m recommendations are found
    
    Notes:
    * Choose the users that have the most total article interactions 
    before choosing those with fewer article interactions.

    * Choose articles with the articles with the most total interactions 
    before choosing those with fewer total interactions. 
   
    '''
    # Your code here
    
    
    # get all similar user ids for the targeted user
    # fetch with the 'neighbors_df'
    similar_uids = list(get_top_sorted_users(user_id).index)
    #print(f"[similar_uids]: \n{similar_uids}")


    # get all article ids of the targeted user
    article_ids_target_user = get_user_articles(user_id)[0]
    # print(f"[article_ids_target_user]:\n {article_ids_target_user} \n")

    
    # a list contain unseen articles to recommend
    recs = []

    for uid in similar_uids:

        # print(f"\n\n\n[number of recs]: {len(recs)}\n")
        if len(recs) == m:
            #print(f"Number of recs reaches threadhold {m}. Enough. Stop")
            break


        #print(f"[similar user id]: {uid}")

        # get this uid's article ids, and arbitrarily shuffle
        article_ids_similar_user = get_user_articles(uid)[0]

        # compute the differences of articles seen between the this user and targeted user
        # subtraction's order matters
        set_diff = list(set(article_ids_similar_user) - set(article_ids_target_user))
        #print(f"[set_diff before sort]:\n {set_diff} \n")
        
        # Sort the set. Determine with highest total interactions metric 
        set_diff = list(df[df.article_id.isin(set_diff)]['article_id'].value_counts().index)
        #print(f"[set_diff after sort]:\n {set_diff} \n")

        # add the differences of article ids to recs [], append only unique (no duplicate)
        for i in set_diff:
            if i not in recs and len(recs) < m:
                recs.append(i)
                #print(f"[id] {i} appended")
    
    
    
    
    rec_names = get_article_names(recs)
    
    return recs, rec_names






In [1311]:
# Quick spot check - don't change this code - just use it to test your functions
rec_ids, rec_names = user_user_recs_part2(20, 10)
print("The top 10 recommendations for user 20 are the following article ids:")
print(rec_ids)
print()
print("The top 10 recommendations for user 20 are the following article names:")
print(rec_names)

The top 10 recommendations for user 20 are the following article ids:
[1330.0, 1427.0, 1364.0, 1170.0, 1162.0, 1304.0, 1351.0, 1160.0, 1354.0, 1368.0]

The top 10 recommendations for user 20 are the following article names:
['insights from new york car accident reports', 'use xgboost, scikit-learn & ibm watson machine learning apis', 'predicting churn with the spss random tree algorithm', 'apache spark lab, part 1: basic concepts', 'analyze energy consumption in buildings', 'gosales transactions for logistic regression model', 'model bike sharing data with spss', 'analyze accident reports on amazon emr spark', 'movie recommender system with spark machine learning', 'putting a human face on machine learning']


`5.` Use your functions from above to correctly fill in the solutions to the dictionary below.  Then test your dictionary against the solution.  Provide the code you need to answer each following the comments below.

In [1312]:
get_top_sorted_users(1).head(1)

neighbor_id  similarity  num_interactions
3933         3933          35                45

In [1316]:
get_top_sorted_users(131).head(10)

neighbor_id  similarity  num_interactions
3870         3870          74               144
3782         3782          39               363
23             23          38               364
203           203          33               160
4459         4459          33               158
98             98          29               170
3764         3764          29               169
49             49          29               147
3697         3697          29               145
242           242          25               148

In [1317]:
### Tests with a dictionary of results

user1_most_sim = 3933 # Find the user that is most similar to user 1 
user131_10th_sim = 242 # Find the 10th most similar user to user 131

In [1318]:
## Dictionary Test Here
sol_5_dict = {
    'The user that is most similar to user 1.': user1_most_sim, 
    'The user that is the 10th most similar to user 131': user131_10th_sim,
}

t.sol_5_test(sol_5_dict)

This all looks good!  Nice job!


6. If we were given a new user, which of the above functions would you be able to use to make recommendations? Explain. Can you think of a better way we might make recommendations? Use the cell below to explain a better method for new users.

A: For new user, code start problem, we can use knowledge base approach, pulling most-interacted (viewed) content and/or trending content. Since the dataset has no timestamp attribute, we might just pull most-interacted content.

7. Using your existing functions, provide the top 10 recommended articles you would provide for the a new user below. You can test your function against our thoughts to make sure we are all on the same page with how we might make a recommendation.

In [1321]:
new_user = 0.0

# What would your recommendations be for this new user '0.0'?  As a new user, they have no observed articles.
# Provide a list of the top 10 article ids you would give to 


new_user_recs = get_top_article_ids(10, df) # Your recommendations here

new_user_recs

[1429.0,
 1330.0,
 1431.0,
 1427.0,
 1364.0,
 1314.0,
 1293.0,
 1170.0,
 1162.0,
 1304.0]

In [1322]:
assert set(new_user_recs) == set([1314.0,1429.0,1293.0,1427.0,1162.0,1364.0,1304.0,1170.0,1431.0,1330.0]), "Oops!  It makes sense that in this case we would want to recommend the most popular articles, because we don't know anything about these users."

print("That's right!  Nice job!")

That's right!  Nice job!


### <a class="anchor" id="Content-Recs">Part IV: Content Based Recommendations (EXTRA - NOT REQUIRED)</a>

Another method we might use to make recommendations is to perform a ranking of the highest ranked articles associated with some term.  You might consider content to be the **doc_body**, **doc_description**, or **doc_full_name**.  There isn't one way to create a content based recommendation, especially considering that each of these columns hold content related information.  

`1.` Use the function body below to create a content based recommender.  Since there isn't one right answer for this recommendation tactic, no test functions are provided.  Feel free to change the function inputs if you decide you want to try a method that requires more input values.  The input values are currently set with one idea in mind that you may use to make content based recommendations.  One additional idea is that you might want to choose the most popular recommendations that meet your 'content criteria', but again, there is a lot of flexibility in how you might make these recommendations.

### This part is NOT REQUIRED to pass this project.  However, you may choose to take this on as an extra way to show off your skills.

In [1446]:
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to /Users/apple/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /Users/apple/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /Users/apple/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/apple/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

https://github.com/gilzero/project_disaster_responses/blob/main/train_classifier.py

https://towardsdatascience.com/the-best-document-similarity-algorithm-in-2020-a-beginners-guide-a01b9ef8cf05

In [1358]:
# constants and reusable objects
url_regex = 'http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
stop_words = stopwords.words("english")
lemmatizer = WordNetLemmatizer()

In [1461]:
df_nlp = df_content.copy()
df_nlp.head(60)

doc_body  \
0   Skip navigation Sign in SearchLoading...\r\n\r...   
1   No Free Hunch Navigation * kaggle.com\r\n\r\n ...   
2   ☰ * Login\r\n * Sign Up\r\n\r\n * Learning Pat...   
3   DATALAYER: HIGH THROUGHPUT, LOW LATENCY AT SCA...   
4   Skip navigation Sign in SearchLoading...\r\n\r...   
5   Compose is all about immediacy. You want a new...   
6   UPGRADING YOUR POSTGRESQL TO 9.5Share on Twitt...   
7   Follow Sign in / Sign up 135 8 * Share\r\n * 1...   
8   * Host\r\n * Competitions\r\n * Datasets\r\n *...   
9   THE GRADIENT FLOW\r\nDATA / TECHNOLOGY / CULTU...   
10  OFFLINE-FIRST IOS APPS WITH SWIFT & PART 1: TH...   
11  Warehousing data from Cloudant to dashDB great...   
12  Skip to main content IBM developerWorks / Deve...   
13  Maureen McElaney Blocked Unblock Follow Follow...   
14  Raj Singh Blocked Unblock Follow Following Dev...   
15  * Home\r\n * Community\r\n * Projects\r\n * Bl...   
16  * Home\r\n * Research\r\n * Partnerships and C...   
17  Enterprise Pricing Articles Sign in Free 30-Da...   
18  Homepage Follow Sign in / Sign up * Home\r\n *...   
19  METRICS MAVEN: MODE D'EMPLOI - FINDING THE MOD...   
20  Homepage Follow Sign in / Sign up Homepage * H...   
21  Raj Singh Blocked Unblock Follow Following Dev...   
22  IMPORTING JSON DOCUMENTS WITH NOSQLIMPORT\r\nG...   
23  This video shows you how to build and query a ...   
24  THE CONVERSATIONAL INTERFACE IS THE NEW PARADI...   
25  Skip navigation Upload Sign in SearchLoading.....   
26  GOOGLE RESEARCH BLOG The latest news from Rese...   
27  Skip navigation Upload Sign in SearchLoading.....   
28  ACCESS DENIED\r\nSadly, your client does not s...   
29  Homepage Follow Sign in / Sign up Homepage * H...   
30  KDNUGGETS\r\nData Mining, Analytics, Big Data,...   
31  Skip navigation Sign in SearchLoading...\r\n\r...   
32  No Free Hunch Navigation * kaggle.com\r\n\r\n ...   
33  Working Vis * \r\n * \r\n\r\n * Home\r\n * Abo...   
34  Homepage Follow Sign in / Sign up Steve Moore ...   
35  Nick Kasten Blocked Unblock Follow Following C...   
36  Jump to navigation\r\n\r\n * Twitter\r\n * Lin...   
37  Homepage IBM Watson Data Lab Follow Sign in / ...   
38  The primary index is the fastest way to retrie...   
39  * R Views\r\n * About this Blog\r\n * Contribu...   
40  Stats and Bots Follow Sign in / Sign up * Home...   
41  TL;DR: It's easy to customise the Mongo shell'...   
42  GETTING STARTED WITH COMPOSE'S SCYLLADB\r\nSha...   
43  ☰ * Login\r\n * Sign Up\r\n\r\n * Learning Pat...   
44  Skip to main content IBM developerWorks / Deve...   
45  Build a custom library for Apache® Spark™ and ...   
46  REAL-TIME Q&A APP WITH RETHINKDB\r\nMatt Colli...   
47  Skip to main content IBM developerWorks / Deve...   
48  Data Science Experience Datasci X * Data Scien...   
49  Compose The Compose logo Articles Sign in Free...   
50  Follow Sign in / Sign up Home About Insight Da...   
51  * Free 7-Day Crash Course\r\n * Blog\r\n * Mas...   
52  * United States\r\n\r\nIBM® * Site map\r\n\r\n...   
53  Homepage Follow Sign in Get started Homepage *...   
54  Jump to navigation\r\n\r\n * Twitter\r\n * Lin...   
55  PREDICT FLIGHT DELAYS WITH APACHE SPARK MLLIB,...   
56  INTRODUCING THE SIMPLE AUTOCOMPLETE SERVICE\r\...   
57  WILL WOLF\r\nDATA SCIENCE THINGS AND THOUGHTS ...   
58  Skip navigation Upload Sign in SearchLoading.....   
59  Homepage PUBLISHED IN AUTONOMOUS AGENTS — #AI ...   

                                      doc_description  \
0   Detect bad readings in real time using Python ...   
1   See the forest, see the trees. Here lies the c...   
2   Here’s this week’s news in Data Science and Bi...   
3   Learn how distributed DBs solve the problem of...   
4   This video demonstrates the power of IBM DataS...   
5            Using Compose's PostgreSQL data browser.   
6   Upgrading your PostgreSQL deployment to versio...   
7   For a company like Slack that strives to be as...   
8   Kaggle is your home for data science. Learn ne... 

In [1431]:
def tokenize(text):
    '''
    private tokenizer to transform each text.
    As a NLP helper function including following tasks:
    - Replace URLs
    - Normalize text
    - Remove punctuation
    - Tokenize words
    - Remove stop words
    - Legmmatize words
    :param text: A message text.
    :return: cleaned tokens extracted from original message text.
    '''

    # print(f"original text: \n {text}")

    # replace urls
    detected_urls = re.findall(url_regex, text)
    for url in detected_urls:
        text = text.replace(url, "urlplaceholder")

    # normalize case and remove punctuation
    text = re.sub(r"[^a-zA-Z0-9]", " ", text.lower())

    # tokenize text
    tokens = word_tokenize(text)

    # lemmatize andremove stop words
    tokens = [lemmatizer.lemmatize(word).strip() for word in tokens if word not in stop_words]

    # in case after normalize/lemmatize, if there is no words, make a dummy element. otherwise StartingVerb breaks
    if len(tokens) < 1:
        tokens = ['none']

    # print(f"tokens: \n {tokens} \n\n")
    return tokens

In [1396]:
# Test tokenized result. 

df_nlp.iloc[0].doc_body

token_doc = tokenize(df_nlp.iloc[0].doc_body)

token_doc

'skip navigation sign searchloading close yeah keep undo closethis video unavailable watch queue queue watch queue queue remove disconnect next video starting stop 1 loading watch queue queue count total find closedemo detect malfunctioning iot sensor streaming analytics ibm analyticsloading unsubscribe ibm analytics cancel unsubscribeworking subscribe subscribed unsubscribe 26kloading loading working add towant watch later sign add video playlist sign share reportneed report video sign report inappropriate content sign transcript statistic add translation 175 view 6like video sign make opinion count sign 7 0don like video sign make opinion count sign 1loading loading transcript interactive transcript could loaded loading loading rating available video rented feature available right please try later published nov 6 2017this video demonstrates streaming analytics application written python running ibm data science experience result analysis displayed map using plotly notebook demonstrat

In [1496]:
# Found index and article id mismatched. Update index with article to make it consistant

df_nlp.index = df_nlp.article_id

In [1637]:
# Cleaning Data. For empty content in body
# There are some rows' doc_body value is NaN
df_nlp.doc_body.isnull()


article_id
0       False
1       False
2       False
3       False
4       False
        ...  
1046    False
1047    False
1048    False
1049    False
1050    False
Name: doc_body, Length: 1051, dtype: bool

In [1634]:
# Update the empty body with 'empty' placeholder
df_nlp.loc[df_nlp[df_nlp.doc_body.isnull()].index, 'doc_body'] = 'empty'

Empty DataFrame
Columns: [doc_body, doc_description, doc_full_name, doc_status, article_id]
Index: []

In [1641]:
# Update the empty desc with 'empty' placeholder
# There are some rows' doc_description value is NaN

df_nlp.loc[df_nlp[df_nlp.doc_description.isnull()].index, 'doc_description'] = 'empty'

In [1566]:
def _compute_article_body_similarity(article_id_1, article_id_2):
    # Compute the consine similarity based on tfidf of body content
    

    doc_a = ' '.join(tokenize(df_nlp.loc[article_id_1].doc_body))
    doc_b = ' '.join(tokenize(df_nlp.loc[article_id_2].doc_body))
    
    

    # combine to a list of documents
    documents = [doc_a, doc_b]

    
    # instanciate a scikitlearn tfidf vecorizer
    vectorizer = TfidfVectorizer()

    # fit transform to get a sparse matrix
    matrix = vectorizer.fit_transform(documents)


    # # UnComment out belows to examing details (term features, tfidf values)
    # ====================================
    # term_features =  vectorizer.get_feature_names_out()

    # # convert to readable array 
    # matrix_array = matrix.toarray()

    # # assemble to a dataframe for explor 
    # tfidf_dataframe = pd.DataFrame(data=matrix_array, columns=term_features)
    # ====================================


    return cosine_similarity(matrix[:1], matrix[1:])[0][0]

In [1567]:
def _compute_article_title_similarity(article_id_1, article_id_2):
    # Compute the consine similarity based on tfidf of title (doc_full_name)
    # think of title tag for seo pagerank
    

    doc_a = ' '.join(tokenize(df_nlp.loc[article_id_1].doc_full_name))
    doc_b = ' '.join(tokenize(df_nlp.loc[article_id_2].doc_full_name))
    
    

    # combine to a list of documents
    documents = [doc_a, doc_b]

    
    # instanciate a scikitlearn tfidf vecorizer
    vectorizer = TfidfVectorizer()

    # fit transform to get a sparse matrix
    matrix = vectorizer.fit_transform(documents)


    # # UnComment out belows to examing details (term features, tfidf values)
    # ====================================
    # term_features =  vectorizer.get_feature_names_out()

    # # convert to readable array 
    # matrix_array = matrix.toarray()

    # # assemble to a dataframe for explor 
    # tfidf_dataframe = pd.DataFrame(data=matrix_array, columns=term_features)
    # ====================================


    return cosine_similarity(matrix[:1], matrix[1:])[0][0]

In [1568]:
def _compute_article_desc_similarity(article_id_1, article_id_2):
    # Compute the consine similarity based on tfidf of desc content (doc_description)
    # think of desc tag for seo pagerank
    

    doc_a = ' '.join(tokenize(df_nlp.loc[article_id_1].doc_description))
    doc_b = ' '.join(tokenize(df_nlp.loc[article_id_2].doc_description))
    
    

    # combine to a list of documents
    documents = [doc_a, doc_b]

    
    # instanciate a scikitlearn tfidf vecorizer
    vectorizer = TfidfVectorizer()

    # fit transform to get a sparse matrix
    matrix = vectorizer.fit_transform(documents)


    # # UnComment out belows to examing details (term features, tfidf values)
    # ====================================
    # term_features =  vectorizer.get_feature_names_out()

    # # convert to readable array 
    # matrix_array = matrix.toarray()

    # # assemble to a dataframe for explor 
    # tfidf_dataframe = pd.DataFrame(data=matrix_array, columns=term_features)
    # ====================================


    return cosine_similarity(matrix[:1], matrix[1:])[0][0]

In [1571]:
def compute_article_similarity(article_id_1, article_id_2):
    # calculate similary for body, title, desc, then combine a single value.
    # think about how google weight. Title tag is very heavy. SEO-wised
    # so having 3 consine similarity values, then do a normailized one. 
    # what's the formular? think of a course, assignments weight x, final exam weight y,
    # then what's total grade.
    # https://www.indeed.com/career-advice/career-development/how-to-calculate-weighted-average
    
    similarity_title = _compute_article_title_similarity(article_id_1,article_id_2)
    similarity_body = _compute_article_body_similarity(article_id_1,article_id_2)
    similarity_desc = _compute_article_desc_similarity(article_id_1,article_id_2)
    
    # a weighted sum caluculation for final score
    
    overall = similarity_title * 0.5 + similarity_body * 0.4 + similarity_desc * 0.1
    
    return overall


In [1576]:
compute_article_similarity(55,50)

0.07122043497708203

In [1794]:
# Find similar articles for a given article
def find_similar_articles(article_id, data=df_nlp):
    
    article_ids_all = data.index
    similarity_dict = {}
    
    for i in article_ids_all:
        # print(f"\n[i]: {i}")
        
        if i == article_id:
            continue

        similarity_score = compute_article_similarity(article_id,i)
        similarity_dict[i] = similarity_score
        # print(f"[similarity_score]: {similarity_score}")
    
    
    similarity_ds = pd.Series(data=similarity_dict.values(), 
                           index=similarity_dict.keys()).sort_values(ascending=False).index
    
    return list(similarity_ds)


In [1795]:
# Calculate in real time, it might take some seconds
# (since it loop to all articles against the targeted article)

relavant_articles = find_similar_articles(420)

relavant_articles

[389,
 993,
 949,
 592,
 714,
 117,
 678,
 942,
 231,
 925,
 15,
 463,
 353,
 835,
 907,
 284,
 977,
 600,
 595,
 997,
 239,
 240,
 411,
 45,
 119,
 473,
 375,
 934,
 560,
 264,
 893,
 515,
 54,
 404,
 849,
 815,
 380,
 486,
 632,
 674,
 424,
 443,
 55,
 9,
 647,
 398,
 161,
 707,
 193,
 948,
 844,
 721,
 681,
 223,
 58,
 961,
 403,
 77,
 874,
 1005,
 762,
 96,
 143,
 112,
 1049,
 938,
 121,
 249,
 235,
 4,
 20,
 341,
 689,
 842,
 809,
 489,
 744,
 141,
 1042,
 872,
 941,
 400,
 594,
 624,
 302,
 318,
 342,
 468,
 272,
 760,
 706,
 280,
 643,
 562,
 47,
 27,
 956,
 499,
 469,
 385,
 7,
 84,
 110,
 326,
 241,
 801,
 728,
 228,
 378,
 683,
 198,
 362,
 138,
 147,
 369,
 593,
 154,
 569,
 847,
 104,
 772,
 447,
 787,
 1028,
 226,
 176,
 578,
 936,
 958,
 13,
 65,
 48,
 334,
 267,
 432,
 634,
 146,
 776,
 293,
 828,
 313,
 44,
 720,
 717,
 846,
 97,
 523,
 120,
 869,
 575,
 691,
 800,
 269,
 1021,
 430,
 262,
 351,
 11,
 12,
 651,
 864,
 62,
 644,
 703,
 729,
 136,
 230,
 773,
 856,
 935,


In [1663]:
# Checking relavancy. They are order by relavancy. 

df_nlp.iloc[relavant_articles].head(20)

# Looks like related to 'Apache Spark'

doc_body  \
article_id                                                      
389         * Home\r\n * Community\r\n * Projects\r\n * Bl...   
993         * Home\r\n * Community\r\n * Projects\r\n * Bl...   
949         * Home\r\n * Community\r\n * Projects\r\n * Bl...   
592         APACHE SPARK ANALYTICSCombine Apache® Spark™ w...   
714         * Home\r\n * Community\r\n * Projects\r\n * Bl...   
117         * Home\r\n * Community\r\n * Projects\r\n * Bl...   
678         { spark .tc } * Community\r\n * Projects\r\n *...   
942         * Home\r\n * Community\r\n * Projects\r\n * Bl...   
231         Skip to main content IBM developerWorks / Deve...   
925         Skip navigation Sign in SearchLoading...\r\n\r...   
15          * Home\r\n * Community\r\n * Projects\r\n * Bl...   
463         Skip navigation Upload Sign in SearchLoading.....   
353         RStudio Blog * Home\r\n\r\n * Subscribe to fee...   
835         Skip to main content IBM developerWorks / Deve...   
907         Skip to main content IBM developerWorks / Deve...   
284         * Home\r\n * Community\r\n * Projects\r\n * Bl...   
977         * Home\r\n * Community\r\n * Projects\r\n * Bl...   
600         Homepage Follow Sign in / Sign up Homepage * H...   
595         Skip to main content IBM developerWorks / Deve...   
997         Skip to main content IBM developerWorks / Deve...   

                                              doc_description  \
article_id                                                      
389         What a difference a version number makes! With...   
993         The Apache Spark website documents the propert...   
949         The Apache Spark SQL component has several sub...   
592         Combine Apache® Spark™ with other cloud servic...   
714         From the big crop of books about Apache Spark™...   
117         This post provides a brief summary of sample c...   
678         Spark SQL Version 1.6 runs queries faster! Tha...   
942         Sean looks back on his first encounter with Sp...   
231         Get faster queries and write less code too. Le...   
925         This video shows you how to use the Spark SQL ...   
15          Early methods to integrate machine learning us...   
463         Learn Why and How To Use Spark for large amoun...   
353         We’re excited today to announce sparklyr, a ne...   
835         How to build SQL Queries in a Scala notebook u...   
907         How to build SQL Queries in a Scala notebook u...   
284         Now that the dust has settled on Apache Spark ...   
977         A handful of talks at the recent Spark Summit ...   
600         In this post I will show you how to use the IB...   
595         Learn how to create a connection to dashDB dat...   
997         One of the best things about Apache Spark is t...   

                                                doc_full_name doc_status  \
article_id                                                                 
389         Apache Spark™ 2.0: Impressive Improvements to ...       Live   
993                  Configuring the Apache Spark SQL Context       Live   
949         Apache Spark SQL Analyzer Resolves Order-by Co...       Live   
592                                    Apache Spark Analytics       Live   
714                     A Survey of Books about Apache Spark™       Live   
117                 Apache Spark™ 2.0: Migrating Applications       Live   
678                   Spark SQL - Rapid Performance Evolution       Live   
942         Interview with Sean Li, New Apache Spark™ Comm...       Live   
231                     Speed your SQL Queries with Spark SQL       Live   
925                Build SQL queries with Apache Spark in DSX       Live   
15          Apache Spark™ 2.0: Extend Structured Streaming...       Live   
463                                            What is Spark?       Live   
353                   sparklyr — R interface for Apache Spark       Live   
835         Build SQL Queries in a Scala notebook u

The above find_similar_articles work well, however it is calculating on the fly, could take time to load, for user experience, real time calculating too expensive. So make a article_article_similary dataframe for look up. (the content-content rec). This can be act as, people view X article also might be interested in Y article based on relavancy.

### Approach:
Make a dataframe, store article-article-similarity.
(Based on the cleaned df_content version dataset: df_nlp)


In [1673]:
# Make a user-user dummy dataframe based on shape of user_article_dict_trimmed
article_article = pd.DataFrame(
                data=np.zeros((len(df_nlp.index),len(df_nlp.index))), 
                index=df_nlp.index, 
                columns=df_nlp.index
            )

article_article

article_id  0     1     2     3     4     5     6     7     8     9     ...  \
article_id                                                              ...   
0            0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
1            0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
2            0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
3            0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
4            0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
...          ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
1046         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
1047         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
1048         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
1049         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
1050         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   

article_id  1041  1042  1043  1044  1045  1046  1047  1048  1049  1050  
article_id                                                              
0            0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1            0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2            0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3            0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4            0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
...          ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
1046         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1047         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1048         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1049         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1050         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[1051 rows x 1051 columns]

In [1692]:
# Iterations. Update the similar score values to article_article

# ATTENTION: THE ITERATION TAKES time, better to track progress


start = time.time()


# Number of articles
len_articles = article_article.shape[0]

# List of index of articles
idx_articles = list(article_article.index)


# Loop thru each article: for each row, loop each columns. 
for i in idx_articles:
    print(f"Processing row {i} of {len_articles-1}.")
    
    for j in idx_articles:
        article_article.loc[i, j] = compute_article_similarity(i, j)
        

end = time.time()
print(end - start)

Processing row 0 of 1050.
Processing row 1 of 1050.
Processing row 2 of 1050.
Processing row 3 of 1050.
Processing row 4 of 1050.
Processing row 5 of 1050.
Processing row 6 of 1050.
Processing row 7 of 1050.
Processing row 8 of 1050.
Processing row 9 of 1050.
Processing row 10 of 1050.
Processing row 11 of 1050.
Processing row 12 of 1050.
Processing row 13 of 1050.
Processing row 14 of 1050.
Processing row 15 of 1050.
Processing row 16 of 1050.
Processing row 17 of 1050.
Processing row 18 of 1050.
Processing row 19 of 1050.
Processing row 20 of 1050.
Processing row 21 of 1050.
Processing row 22 of 1050.
Processing row 23 of 1050.
Processing row 24 of 1050.
Processing row 25 of 1050.
Processing row 26 of 1050.
Processing row 27 of 1050.
Processing row 28 of 1050.
Processing row 29 of 1050.
Processing row 30 of 1050.
Processing row 31 of 1050.
Processing row 32 of 1050.
Processing row 33 of 1050.
Processing row 34 of 1050.
Processing row 35 of 1050.
Processing row 36 of 1050.
Processing 

Processing row 297 of 1050.
Processing row 298 of 1050.
Processing row 299 of 1050.
Processing row 300 of 1050.
Processing row 301 of 1050.
Processing row 302 of 1050.
Processing row 303 of 1050.
Processing row 304 of 1050.
Processing row 305 of 1050.
Processing row 306 of 1050.
Processing row 307 of 1050.
Processing row 308 of 1050.
Processing row 309 of 1050.
Processing row 310 of 1050.
Processing row 311 of 1050.
Processing row 312 of 1050.
Processing row 313 of 1050.
Processing row 314 of 1050.
Processing row 315 of 1050.
Processing row 316 of 1050.
Processing row 317 of 1050.
Processing row 318 of 1050.
Processing row 319 of 1050.
Processing row 320 of 1050.
Processing row 321 of 1050.
Processing row 322 of 1050.
Processing row 323 of 1050.
Processing row 324 of 1050.
Processing row 325 of 1050.
Processing row 326 of 1050.
Processing row 327 of 1050.
Processing row 328 of 1050.
Processing row 329 of 1050.
Processing row 330 of 1050.
Processing row 331 of 1050.
Processing row 332 o

Processing row 590 of 1050.
Processing row 591 of 1050.
Processing row 592 of 1050.
Processing row 593 of 1050.
Processing row 594 of 1050.
Processing row 595 of 1050.
Processing row 596 of 1050.
Processing row 597 of 1050.
Processing row 598 of 1050.
Processing row 599 of 1050.
Processing row 600 of 1050.
Processing row 601 of 1050.
Processing row 602 of 1050.
Processing row 603 of 1050.
Processing row 604 of 1050.
Processing row 605 of 1050.
Processing row 606 of 1050.
Processing row 607 of 1050.
Processing row 608 of 1050.
Processing row 609 of 1050.
Processing row 610 of 1050.
Processing row 611 of 1050.
Processing row 612 of 1050.
Processing row 613 of 1050.
Processing row 614 of 1050.
Processing row 615 of 1050.
Processing row 616 of 1050.
Processing row 617 of 1050.
Processing row 618 of 1050.
Processing row 619 of 1050.
Processing row 620 of 1050.
Processing row 621 of 1050.
Processing row 622 of 1050.
Processing row 623 of 1050.
Processing row 624 of 1050.
Processing row 625 o

Processing row 883 of 1050.
Processing row 884 of 1050.
Processing row 885 of 1050.
Processing row 886 of 1050.
Processing row 887 of 1050.
Processing row 888 of 1050.
Processing row 889 of 1050.
Processing row 890 of 1050.
Processing row 891 of 1050.
Processing row 892 of 1050.
Processing row 893 of 1050.
Processing row 894 of 1050.
Processing row 895 of 1050.
Processing row 896 of 1050.
Processing row 897 of 1050.
Processing row 898 of 1050.
Processing row 899 of 1050.
Processing row 900 of 1050.
Processing row 901 of 1050.
Processing row 902 of 1050.
Processing row 903 of 1050.
Processing row 904 of 1050.
Processing row 905 of 1050.
Processing row 906 of 1050.
Processing row 907 of 1050.
Processing row 908 of 1050.
Processing row 909 of 1050.
Processing row 910 of 1050.
Processing row 911 of 1050.
Processing row 912 of 1050.
Processing row 913 of 1050.
Processing row 914 of 1050.
Processing row 915 of 1050.
Processing row 916 of 1050.
Processing row 917 of 1050.
Processing row 918 o

In [1694]:
# Check updated article_article with calculated results. 

article_article

article_id      0         1         2         3         4         5     \
article_id                                                               
0           1.000000  0.032471  0.072460  0.020439  0.259333  0.033410   
1           0.032471  1.000000  0.186967  0.028671  0.093232  0.098205   
2           0.072460  0.186967  1.000000  0.053018  0.172458  0.106455   
3           0.020439  0.028671  0.053018  1.000000  0.027609  0.037732   
4           0.259333  0.093232  0.172458  0.027609  1.000000  0.095990   
...              ...       ...       ...       ...       ...       ...   
1046        0.003492  0.014128  0.009416  0.012851  0.006955  0.019682   
1047        0.004095  0.019350  0.013930  0.011922  0.003037  0.010529   
1048        0.028436  0.056505  0.104403  0.010652  0.038724  0.026182   
1049        0.000000  0.010927  0.018914  0.009425  0.072108  0.010935   
1050        0.078722  0.090637  0.160396  0.036531  0.138237  0.030091   

article_id      6         7         8         9     ...      1041      1042  \
article_id                                          ...                       
0           0.016156  0.051355  0.046419  0.047824  ...  0.063201  0.051671   
1           0.035012  0.155317  0.206362  0.103742  ...  0.104562  0.152361   
2           0.041646  0.204603  0.339824  0.153497  ...  0.149979  0.208527   
3           0.084109  0.048761  0.031276  0.049991  ...  0.027299  0.026927   
4           0.026990  0.154261  0.135846  0.152328  ...  0.122135  0.141956   
...              ...       ...       ...       ...  ...       ...       ...   
1046        0.009555  0.018015  0.004764  0.006668  ...  0.013567  0.003230   
1047        0.002925  0.006396  0.007393  0.017164  ...  0.007295  0.002512   
1048        0.017823  0.050228  0.041883  0.069507  ...  0.074214  0.024484   
1049        0.000000  0.012977  0.014043  0.066733  ...  0.005548  0.069338   
1050        0.038946  0.130746  0.084648  0.095793  ...  0.188039  0.113751   

article_id      1043      1044      1045      1046      1047      1048  \
article_id                                                               
0           0.143709  0.020163  0.010646  0.003492  0.004095  0.028436   
1           0.093032  0.030101  0.082490  0.014128  0.019350  0.056505   
2           0.139453  0.036274  0.009098  0.009416  0.013930  0.104403   
3           0.033499  0.010938  0.036107  0.012851  0.011922  0.010652   
4           0.137278  0.024670  0.007071  0.006955  0.003037  0.038724   
...              ...       ...       ...       ...       ...       ...   
1046        0.003622  0.014652  0.005776  1.000000  0.000000  0.001758   
1047        0.005704  0.001240  0.006689  0.000000  1.000000  0.058308   
1048        0.034961  0.020447  0.015387  0.001758  0.058308  1.000000   
1049        0.068079  0.005126  0.000000  0.000000  0.000000  0.000000   
1050        0.140985  0.025935  0.023651  0.005407  0.006413  0.114996   

article_id      1049      1050  
article_id                      
0           0.000000  0.078722  
1           0.010927  0.090637  
2           0.018914  0.160396  
3           0.009425  0.036531  
4           0.072108  0.138237  
...              ...       ...  
1046        0.000000  0.005407  
1047        0.000000  0.006413  
1048        0.000000  0.114996  
1049        1.000000  0.000000  
1050        0.000000  1.000000  

[1051 rows x 1051 columns]

In [1695]:
# Export to pickle

article_article.to_pickle('article_article_similarity_df.pkl')

In [1777]:
# Instead of finding in real time. Use a query method from pre-calculated 
# article_article df. Save time.

def loopup_similar_articles(article_id, data=article_article, n=20):
    
    # input: n - number of top similar to return
    
    ids = list(data.loc[article_id][data.loc[article_id].index != article_id].sort_values(ascending=False).head(n).index)
    names = list(df_nlp.loc[ids].doc_full_name.values)
    
    
    return ids, names


In [1785]:
# Test result: look up most relavant articles for 455th article

relavant_for_455th_article = loopup_similar_articles(455, n=30)
relavant_for_455th_article

# Revealing that they are about on 'machine learning'.

([800,
  1035,
  313,
  805,
  444,
  721,
  260,
  967,
  122,
  96,
  384,
  809,
  124,
  723,
  234,
  892,
  54,
  253,
  812,
  861,
  412,
  732,
  871,
  74,
  221,
  89,
  479,
  616,
  500,
  567],
 ['Machine Learning for the Enterprise',
  'Machine Learning for the Enterprise.',
  'What is machine learning?',
  'Machine Learning for everyone',
  'Declarative Machine Learning',
  'The power of machine learning in Spark',
  'The Machine Learning Database',
  'ML Algorithm != Learning Machine',
  'Watson Machine Learning for Developers',
  'Improving quality of life with Spark-empowered machine learning',
  'Continuous Learning on Watson',
  'Use the Machine Learning Library',
  'Python Machine Learning: Scikit-Learn Tutorial',
  '10 Essential Algorithms For Machine Learning Engineers',
  '3 Scenarios for Machine Learning on Multicloud',
  'Breaking the 80/20 rule: How data catalogs transform data scientists’ productivity',
  '8 ways to turn data into value with Apache Spark ma

### Note on datasets inconsistency: 

Since that orginal df_content / df_nlp (cleaned version) only contain article from \[0 - 1050\].

However, df (the interaction) dataset showing more article ids that beyone 1050. e,g 1429.

So these two dataset's article are not up to dated.  
The given df_content is 'late' (not catch up), which means: **articles that are in df (interaction) dataset might 
not be found in df_content dataset**. 

Therefore, due to this inconsistency, we will not be able to find content details for some articles mentioned in df.

----


### Other ideas to explore: 
- Based on the relavant article for a given article, can you get common words on them. Extract topic perhaps
- What are the top contents? (that also in df_content/df_nlp)
- What are the common words / terms of those top content? word bag counter
- Can these words/terms represent trending topic?
- Make a list of most trending topics


In [1786]:
# Extract topics from relavant content.
# Try extract common terms represented in top relavant articles. 
# e.g the relavant results showing they all contain 'machine learning'. 
# Approach with intersection of tokens for those relavant documents.
# bag of words, 1-gram, 2-gram, 3-gram? counter? then intersections 

def extract_topics_from_relavant_articles(ids, data=df_nlp):
    
    return []



In [1808]:
target_id = 455
relevant_articles = loopup_similar_articles(target_id, n=30)

r_ids = relevant_articles[0]
r_ids

# concat tokens: for an article, combine token of title, desc, body all together to represent its tokened text
test_id = 800

title = df_nlp.loc[test_id].doc_full_name
desc = df_nlp.loc[test_id].doc_description
body = df_nlp.loc[test_id].doc_body

tokened_text = tokenize(title) + tokenize(desc) + tokenize(body)
tokened_text





['machine',
 'learning',
 'enterprise',
 'last',
 'blog',
 'business',
 'differentiation',
 'machine',
 'learning',
 'introduced',
 'described',
 'concept',
 'machine',
 'learning',
 'traced',
 'origin',
 'computer',
 'science',
 'project',
 'watson',
 'show',
 'skip',
 'contentdinesh',
 'nirmal',
 'blog',
 'blog',
 'big',
 'data',
 'analytics',
 'digital',
 'technology',
 'impacting',
 'business',
 'data',
 'driven',
 'economy',
 'menu',
 'welcome',
 'twitter',
 'machine',
 'learning',
 'enterprise',
 'last',
 'blog',
 'business',
 'differentiation',
 'machine',
 'learning',
 'introduced',
 'described',
 'concept',
 'machine',
 'learning',
 'traced',
 'origin',
 'computer',
 'science',
 'project',
 'watson',
 'showcasing',
 'winning',
 'jeopardy',
 'tv',
 'quiz',
 'show',
 'real',
 'world',
 'use',
 'across',
 'numerous',
 'industry',
 'including',
 'health',
 'care',
 'concluded',
 'machine',
 'learning',
 'potential',
 'help',
 'make',
 'world',
 'better',
 'safer',
 'place',
 'howe

In [1809]:
from nltk.util import pad_sequence
from nltk.util import bigrams
from nltk.util import ngrams
from nltk.util import everygrams
from nltk.lm.preprocessing import pad_both_ends
from nltk.lm.preprocessing import flatten

In [1823]:
# two grams

[f"{x[0]} {x[1]}" for x in list(bigrams(tokened_text))]

['machine learning',
 'learning enterprise',
 'enterprise last',
 'last blog',
 'blog business',
 'business differentiation',
 'differentiation machine',
 'machine learning',
 'learning introduced',
 'introduced described',
 'described concept',
 'concept machine',
 'machine learning',
 'learning traced',
 'traced origin',
 'origin computer',
 'computer science',
 'science project',
 'project watson',
 'watson show',
 'show skip',
 'skip contentdinesh',
 'contentdinesh nirmal',
 'nirmal blog',
 'blog blog',
 'blog big',
 'big data',
 'data analytics',
 'analytics digital',
 'digital technology',
 'technology impacting',
 'impacting business',
 'business data',
 'data driven',
 'driven economy',
 'economy menu',
 'menu welcome',
 'welcome twitter',
 'twitter machine',
 'machine learning',
 'learning enterprise',
 'enterprise last',
 'last blog',
 'blog business',
 'business differentiation',
 'differentiation machine',
 'machine learning',
 'learning introduced',
 'introduced described'

In [1826]:
# 3-gram

list(ngrams(tokened_text, n=3))

[f"{x[0]} {x[1]} {x[2]}" for x in list(ngrams(tokened_text, n=3))]

['machine learning enterprise',
 'learning enterprise last',
 'enterprise last blog',
 'last blog business',
 'blog business differentiation',
 'business differentiation machine',
 'differentiation machine learning',
 'machine learning introduced',
 'learning introduced described',
 'introduced described concept',
 'described concept machine',
 'concept machine learning',
 'machine learning traced',
 'learning traced origin',
 'traced origin computer',
 'origin computer science',
 'computer science project',
 'science project watson',
 'project watson show',
 'watson show skip',
 'show skip contentdinesh',
 'skip contentdinesh nirmal',
 'contentdinesh nirmal blog',
 'nirmal blog blog',
 'blog blog big',
 'blog big data',
 'big data analytics',
 'data analytics digital',
 'analytics digital technology',
 'digital technology impacting',
 'technology impacting business',
 'impacting business data',
 'business data driven',
 'data driven economy',
 'driven economy menu',
 'economy menu wel

In [1832]:
# helper function: get two grams

def get_ngrams_2_by_article_id(id):
    title = df_nlp.loc[id].doc_full_name
    desc = df_nlp.loc[id].doc_description
    body = df_nlp.loc[id].doc_body
    
    # a list of tokens
    tokened_text = tokenize(title) + tokenize(desc) + tokenize(body)
    
    # ngrams 2
    ngrams = [f"{x[0]} {x[1]}" for x in list(bigrams(tokened_text))]
    
    return ngrams

In [1844]:
a = get_ngrams_2_by_article_id(949)
a

['apache spark',
 'spark sql',
 'sql analyzer',
 'analyzer resolve',
 'resolve order',
 'order column',
 'column apache',
 'apache spark',
 'spark sql',
 'sql component',
 'component several',
 'several sub',
 'sub component',
 'component including',
 'including analyzer',
 'analyzer play',
 'play important',
 'important role',
 'role making',
 'making sure',
 'sure logical',
 'logical plan',
 'plan fully',
 'fully resolved',
 'resolved end',
 'end analysis',
 'analysis phase',
 'phase analyzer',
 'analyzer take',
 'take parsed',
 'parsed logical',
 'logical plan',
 'plan input',
 'input make',
 'make sure',
 'sure table',
 'table reference',
 'reference attribute',
 'attribute column',
 'column reference',
 'reference function',
 'function reference',
 'reference resolved',
 'resolved looking',
 'looking metadata',
 'metadata catalog',
 'catalog work',
 'work applying',
 'applying set',
 'set rule',
 'rule logical',
 'logical plan',
 'plan transforming',
 'transforming stage',
 'stage

In [1845]:
b = get_ngrams_2_by_article_id(592)
b

['apache spark',
 'spark analytics',
 'analytics combine',
 'combine apache',
 'apache spark',
 'spark cloud',
 'cloud service',
 'service speed',
 'speed analysis',
 'analysis reveal',
 'reveal insight',
 'insight apache',
 'apache spark',
 'spark analyticscombine',
 'analyticscombine apache',
 'apache spark',
 'spark cloud',
 'cloud service',
 'service speed',
 'speed analysis',
 'analysis revealinsights',
 'revealinsights whether',
 'whether want',
 'want track',
 'track conversation',
 'conversation trending',
 'trending twitter',
 'twitter predictfuture',
 'predictfuture event',
 'event based',
 'based trove',
 'trove existing',
 'existing data',
 'data lighting',
 'lighting fast',
 'fast processingpower',
 'processingpower apache',
 'apache spark',
 'spark help',
 'help crunch',
 'crunch large',
 'large data',
 'data set',
 'set fast',
 'fast show',
 'show youhow',
 'youhow harness',
 'harness power',
 'power potential',
 'potential apache',
 'apache spark',
 'spark engine',
 'en

In [1847]:
c = get_ngrams_2_by_article_id(15)
c

['apache spark',
 'spark 2',
 '2 0',
 '0 extend',
 'extend structured',
 'structured streaming',
 'streaming spark',
 'spark ml',
 'ml early',
 'early method',
 'method integrate',
 'integrate machine',
 'machine learning',
 'learning using',
 'using naive',
 'naive bayes',
 'bayes custom',
 'custom sink',
 'sink home',
 'home community',
 'community project',
 'project blog',
 'blog advisory',
 'advisory council',
 'council resource',
 'resource code',
 'code contribution',
 'contribution university',
 'university ibm',
 'ibm design',
 'design apache',
 'apache systemml',
 'systemml apache',
 'apache spark',
 'spark spark',
 'spark tc',
 'tc community',
 'community project',
 'project blog',
 'blog advisory',
 'advisory council',
 'council resource',
 'resource code',
 'code contribution',
 'contribution university',
 'university ibm',
 'ibm design',
 'design apache',
 'apache systemml',
 'systemml apache',
 'apache spark',
 'spark streaming',
 'streaming extend',
 'extend structured'

In [1852]:
a[0:10]

np.intersect1d(a[0:10], b[0:10])

array(['apache spark'], dtype='<U17')

In [1862]:
ab = np.intersect1d(a, b)
ab

abc = np.intersect1d(ab, np.array(c)
abc

array(['apache software', 'apache spark', 'open source',
       'software foundation', 'spark spark', 'spark trademark'],
      dtype='<U51')